corehq/apps/app_manager/tests/test_util.py

```diff
@@ -0,0 +1,98 @@
+        })
+
+    def test_get_casedb_schema_with_form(self):
+        app = self.make_app()
+        self.add_form(app, "village")
+        schema = util.get_casedb_schema(app)
+        self.assertEqual(len(schema["subsets"]), 1, schema["subsets"])
+        self.assert_has_kv_pairs(schema["subsets"][0], {
+            'id': 'village',
+            'key': '@case_type',
+            'structure': {'name': {}},
+            'related': None,
+        })
+
+    def test_get_casedb_schema_with_related_case_types(self):
+        app = self.make_app()
+        self.add_form(app, "family")
+        village = self.add_form(app, "village")
+        village.actions.subcases.append(OpenSubCaseAction(
+            case_type='family',
+            reference_id='parent'
+        ))
+        schema = util.get_casedb_schema(app)
+        subsets = {s["id"]: s for s in schema["subsets"]}
+        self.assertEqual(subsets["village"]["related"], None)
+        self.assertDictEqual(subsets["family"]["related"], {"parent": "village"})
+
+    def test_get_session_schema_for_module_with_no_case_type(self):
+        app = self.make_app()
+        form = self.add_form(app)
+        schema = util.get_session_schema(form)
+        self.assert_has_kv_pairs(schema, {
+            "id": "commcaresession",
+            "uri": "jr://instance/session",
+            "name": "Session",
+            "path": "/session/data",
+        })
+        assert "case_id" not in schema["structure"], schema["structure"]
+
+    def test_get_session_schema_for_simple_module_with_case(self):
+        app = self.make_app()
+        form = self.add_form(app, "village")
+        schema = util.get_session_schema(form)
+        self.assertDictEqual(schema["structure"]["case_id"], {
+            "reference": {
+                "source": "casedb",
+                "subset": "village",
+                "key": "@case_id",
+            },
+        })
+
+    # -- helpers --
+
+    def assert_has_kv_pairs(self, test_dict, expected_dict):
+        """Assert that test_dict contains all key/value pairs in expected_dict
+
+        Key/value pairs in `test_dict` but not present in
+        `expected_dict` will be ignored.
+        """
+        for key, value in expected_dict.items():
+            self.assertEqual(test_dict[key], value)
+
+    def add_form(self, app, case_type=None, module_id=None):
+        if module_id is None:
+            module_id = len(app.modules)
+            m = app.add_module(Module.new_module('Module{}'.format(module_id), lang='en'))
+            if case_type:
+                m.case_type = case_type
+        form = app.new_form(module_id, 'form {}'.format(case_type), lang='en')
+        if case_type:
+            form.actions.open_case = OpenCaseAction(name_path="/data/question1", external_id=None)
+            form.actions.open_case.condition.type = 'always'
+        return form
+
+    def make_app(self):
+        # empty domain means util.is_usercase_in_use(domain) -> False
```

Comment:
I had initially tried that, but it turned out to be hard because we'd need to patch `corehq.apps.app_manager.models.is_usercase_in_use` (not where the function is defined) in addition to `corehq.apps.app_manager.util.is_usercase_in_use`, which seemed even more fragile. How many other places is that function imported that would need to be patched as well?

I agree, this does have a smell. Do you have any other suggestions for how to clean it up?